### PR TITLE
Add Terraform builder

### PR DESCRIPTION
**Pull request by Thomas-Roper**
Terraform cloud builder with examples of local and GCS backends.

### CODE REVIEW

**Comment by Thomas-Roper** [comment]
Rebased PR from master

**Comment by Thomas-Roper** [comment]
@ImJasonH do you have questions or concerns about this builder? Anything I can do to help with a review?

-Tom

**Comment by Thomas-Roper** [comment]
@ImJasonH @nof20 Anything I can do to help out with this PR?

**Comment by polleyg** [comment]
It would be great to see a Terraform builder as an option. Any chance of getting this over the line?

File: terraform/README.markdown [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# [Terraform](https://www.terraform.io/docs) cloud builder
+
+Terraform builder requested [here](https://github.com/GoogleCloudPlatform/cloud-builders/issues/233)
```
Comment (ImJasonH) [inline terraform/README.markdown]: I'm not sure this link adds much to the docs.

File: terraform/README.markdown [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+## Terraform cloud builder
+This builder can be used to run the terraform tool in the GCE.
```
Comment (ImJasonH) [inline terraform/README.markdown]: Can you add a brief description of what Terraform can be used for?

File: terraform/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,17 @@
+FROM gcr.io/cloud-builders/gsutil
```
Comment (ImJasonH) [inline terraform/Dockerfile]: The gsutil builder is just the gcloud builder with an overridden entrypoint. If you're going to override the entrypoint yourself you might as well make this `FROM gcr.io/cloud-builders/gcloud`

File: terraform/examples/gcs_backend/cloudbuild.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# In this directory, run the following command to build this builder.
+# $ gcloud container builds submit . --config=cloudbuild.yaml
+steps:
+- name: 'gcr.io/${PROJECT_ID}/terraform' # Remove this step after the first run. TODO: Track down optional steps documentation
```
Comment (ImJasonH) [inline terraform/examples/gcs_backend/cloudbuild.yaml]: Optional steps are not supported at this time. You can either:

1. Document that the bucket must exist as a prerequisite to the build
1. Allow this step to fail by using this terrible hack:

```
- name: gcr.io/${PROJECT_ID}/terraform
  entrypoint: 'bash'
  args: ['-c', 'gsutil mb -l us-east1 gs://{$_BUCKET} || true']
```

**Comment by Thomas-Roper** [comment]
@ImJasonH I think I answered your feedback; pls lemme know if there's anything else you can see. 
Thanks,
-Tom

**Comment by Thomas-Roper** [comment]
@ImJasonH is there anything I can do to help this along? @polleyg have you been able to give this builder a try? Do you have any feedback?

**Comment by ImJasonH** [comment]
@Thomas-Roper All looks good to me. Thanks! 🎉

**Pull request merged**